# 📊 BugFix Dataset Overview & Exploratory Data Analysis

This notebook provides a detailed inspection, metrics, and visual breakdown of the collected bug-fix dataset stored in **`dataset_enriched_v3.csv`**.

---

## 1. Load Dataset & Overall Metrics

In [2]:
import os
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

# Set styling
sns.set_theme(style="whitegrid", palette="muted")
plt.rcParams["figure.figsize"] = (10, 6)
plt.rcParams["font.size"] = 11

# Load CSV
csv_path = "dataset_enriched_v3.csv"
df = pd.read_csv(csv_path)
file_size_kb = os.path.getsize(csv_path) / 1024

print("=== DATASET OVERVIEW ===")
print(f"File Path         : {csv_path}")
print(f"File Size         : {file_size_kb:.2f} KB")
print(f"Total Bug Samples : {len(df)}")
print(f"Total Features    : {len(df.columns)}")
print(f"Unique Repos      : {df['repository'].nunique()}")

df.info()

=== DATASET OVERVIEW ===
File Path         : dataset_enriched_v3.csv
File Size         : 3801.75 KB
Total Bug Samples : 478
Total Features    : 26
Unique Repos      : 290
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 478 entries, 0 to 477
Data columns (total 26 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   repository                478 non-null    object 
 1   commit_hash               478 non-null    object 
 2   file_path                 478 non-null    object 
 3   language                  478 non-null    object 
 4   bug_type                  478 non-null    object 
 5   original_commit_message   478 non-null    object 
 6   buggy_code                478 non-null    object 
 7   fixed_code                478 non-null    object 
 8   buggy_ast_vector          478 non-null    object 
 9   fixed_ast_vector          478 non-null    object 
 10  ast_tier                  478 non-null    object 
 11  bug_

In [ ]:
# Sample view of records
display(df[["repository", "language", "bug_type", "ast_tier", "severity", "gemini_confidence"]].head(10))

## 2. Distribution of Bug Categories (`bug_type`)
Inspecting the primary taxonomy of bugs mined across GitHub repositories.

In [ ]:
bug_counts = df["bug_type"].value_counts()
bug_pct = df["bug_type"].value_counts(normalize=True) * 100
bug_summary = pd.DataFrame({"Count": bug_counts, "Percentage (%)": bug_pct.round(2)})
display(bug_summary)

plt.figure(figsize=(12, 6))
ax = sns.barplot(x=bug_counts.values, y=bug_counts.index, hue=bug_counts.index, palette="viridis", legend=False)
plt.title("Distribution of Bug Types in Dataset", fontsize=14, fontweight="bold", pad=15)
plt.xlabel("Number of Collected Samples", fontsize=12)
plt.ylabel("Bug Category", fontsize=12)
for i, v in enumerate(bug_counts.values):
    ax.text(v + 0.3, i, f"{v} ({bug_pct.iloc[i]:.1f}%)", va="center", fontweight="bold")
plt.tight_layout()
plt.show()

In [3]:
data=pd.read_csv(r'C:\Users\YATIRAJ\OneDrive\Desktop\claude3\binary_bug_detection_dataset.csv')
data

,sample_id,pair_id,is_bug,label_name,code,ast_vector,language,bug_type,severity,repository,commit_hash,file_path,ast_tier,bug_description,root_cause,fix_description,gemini_confidence,gemini_validation_status
0,1,0,1,buggy,const Blogs = () => {\n const posts = useSele...,"[0.0, 0.0, 0.16982648, 0.0, 0.0, 0.0, 0.0, 0.1...",javascript,logic_error,medium,thecodechaser/portfolio-frontend,835e203908a353b83e6a5fd42666dd1756086a64,src/pages/Blogs.js,tier_1_direct,"When fetchPostsApi returns no posts, the UI sh...",Component did not handle the empty posts state...,Added a conditional check: if posts.length ===...,0.90,validated
1,2,0,0,clean,const Blogs = () => {\n const posts = useSele...,"[0.0, 0.0, 0.15248875, 0.0, 0.0, 0.0, 0.0, 0.1...",javascript,logic_error,medium,thecodechaser/portfolio-frontend,835e203908a353b83e6a5fd42666dd1756086a64,src/pages/Blogs.js,tier_1_direct,"When fetchPostsApi returns no posts, the UI sh...",Component did not handle the empty posts state...,Added a conditional check: if posts.length ===...,0.90,validated
2,3,1,1,buggy,export const MilestonesPage = () => {\n const...,"[0.0, 0.0, 0.04768815, 0.0, 0.0, 0.0, 0.0, 0.1...",typescript,state_management,medium,reportportal/service-ui,2d213871549f7ea57700ea5afc0003488339d92d,app/src/pages/inside/testPlansPage/milestonesP...,tier_1_direct,MilestonesPage dispatched getMilestonesAction ...,The component fetched milestones without check...,Added an isAuthorized selector and included it...,0.90,validated
3,4,1,0,clean,export const MilestonesPage = () => {\n const...,"[0.0, 0.0, 0.04570412, 0.0, 0.0, 0.0, 0.0, 0.1...",typescript,state_management,medium,reportportal/service-ui,2d213871549f7ea57700ea5afc0003488339d92d,app/src/pages/inside/testPlansPage/milestonesP...,tier_1_direct,MilestonesPage dispatched getMilestonesAction ...,The component fetched milestones without check...,Added an isAuthorized selector and included it...,0.90,validated
4,5,2,1,buggy,const publish = async () => {\n try {\n ...,"[0.0, 0.0, 0.03973949, 0.0, 0.0, 0.0, 0.0, 0.1...",javascript,exception_handling,medium,ipfs/ipfs-webui,5cdb3013938b87fad6510b70ca6139b3ce4d4fe9,src/files/modals/publish-modal/PublishModal.js,tier_1_direct,"When publishing to IPNS failed, the error was ...",The catch block did not reset UI state (start ...,Added setError(null) before the async call to ...,0.95,validated
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
951,952,475,0,clean,const App = () => {\n\n const {language} = ...,"[0.0, 0.0, 0.09117531, 0.0, 0.0, 0.0, 0.0, 0.0...",javascript,logic_error,medium,Vishal023/Vusic,46f26b7774678b9c65535629fd5ddd7bd36f1dbc,src/app/App.jsx,tier_1_direct,When the selected language string includes 'an...,Missing condition to handle language values co...,Added a check for `language.includes('any')` a...,0.90,validated
952,953,476,1,buggy,"function Home() {\n\n\n const [screenSize, ...","[0.03831318, 0.0, 0.07662636, 0.0, 0.0, 0.0, 0...",javascript,state_management,medium,Vishal023/Vusic,46f26b7774678b9c65535629fd5ddd7bd36f1dbc,src/components/Pages/Home.jsx,tier_1_direct,"After navigating to a different route (e.g., /...",The component read `window.location.pathname` ...,Introduced a `Page` state variable and a `useE...,0.90,validated
953,954,476,0,clean,"function Home() {\n\n\n const [screenSize, ...","[0.03331485, 0.0, 0.08328713, 0.0, 0.0, 0.0, 0...",javascript,state_management,medium,Vishal023/Vusic,46f26b7774678b9c65535629fd5ddd7bd36f1dbc,src/components/Pages/Home.jsx,tier_1_direct,"After navigating to a different route (e.g., /...",The component read `window.location.pathname` ...,Introduced a `Page` state variable and a `useE...,0.90,validated
954,955,477,1,buggy,const SearchBar = () => {\n const [isSearch...,"[0.0, 0.0, 0.08209322, 0.0, 0.0, 0.0, 0.0, 0.1...",javascript,state_management,medium,Vishal023/Vusic,46f26b7774678b9c65535629fd5ddd7bd36f1dbc,src/components/fragment/SearchBar.jsx,tier_1_direct,Search terms with uppercase letters were not m

## 3. Programming Languages & AST Reliability Tiers
Checking programming language breakdown (JavaScript vs TypeScript) and AST extraction reliability tiers.

In [ ]:
fig, axes = plt.subplots(1, 2, figsize=(14, 5))

# Language breakdown
lang_counts = df["language"].value_counts()
axes[0].pie(lang_counts, labels=lang_counts.index, autopct="%1.1f%%", startangle=140, colors=["#3498db", "#e74c3c"], explode=[0.05]*len(lang_counts))
axes[0].set_title("Language Breakdown", fontsize=13, fontweight="bold")

# AST Tier breakdown
ast_counts = df["ast_tier"].value_counts()
sns.barplot(x=ast_counts.index, y=ast_counts.values, ax=axes[1], hue=ast_counts.index, palette="mako", legend=False)
axes[1].set_title("AST Vector Reliability Tier", fontsize=13, fontweight="bold")
axes[1].set_ylabel("Sample Count")
for i, v in enumerate(ast_counts.values):
    axes[1].text(i, v + 0.5, str(v), ha="center", fontweight="bold")

plt.tight_layout()
plt.show()

## 4. LLM Semantic Judge & Confidence Analysis
Analyzing model confidence scores and validation status from Nemotron / Gemini LLM evaluation.

In [ ]:
fig, axes = plt.subplots(1, 2, figsize=(14, 5))

# Confidence distribution
sns.histplot(df["gemini_confidence"], kde=True, ax=axes[0], color="purple", bins=10)
axes[0].set_title("LLM Validation Confidence Score Distribution", fontsize=13, fontweight="bold")
axes[0].set_xlabel("Confidence Score")
axes[0].set_ylabel("Count")

# Severity Breakdown
sev_counts = df["severity"].value_counts()
sns.barplot(x=sev_counts.index, y=sev_counts.values, ax=axes[1], hue=sev_counts.index, palette="rocket", legend=False)
axes[1].set_title("Bug Severity Distribution", fontsize=13, fontweight="bold")
axes[1].set_ylabel("Count")
for i, v in enumerate(sev_counts.values):
    axes[1].text(i, v + 0.5, str(v), ha="center", fontweight="bold")

plt.tight_layout()
plt.show()

## 5. Top Contributing Repositories

In [ ]:
top_repos = df["repository"].value_counts().head(10)
plt.figure(figsize=(12, 5))
sns.barplot(x=top_repos.values, y=top_repos.index, hue=top_repos.index, palette="coolwarm", legend=False)
plt.title("Top 10 Source Repositories in Dataset", fontsize=13, fontweight="bold")
plt.xlabel("Number of Bug Fixes Mined")
plt.ylabel("Repository")
for i, v in enumerate(top_repos.values):
    plt.text(v + 0.05, i, str(v), va="center", fontweight="bold")
plt.tight_layout()
plt.show()

## 6. Summary & Insights

### Data Analysis Key Findings
- **Total Dataset Volume**: **91 validated bug samples** spanning **63 unique open-source GitHub repositories** (~707 KB CSV size).
- **Dominant Bug Categories**: **State Management** (27 samples, 29.7%), **Logic Errors** (17 samples, 18.7%), **Null Pointer** (12 samples, 13.2%), and **Async Errors** (10 samples, 11.0%).
- **Language Breakdown**: **JavaScript** accounts for **92.3%** (84 samples), while **TypeScript** makes up **7.7%** (7 samples).
- **High Quality & Confidence**: Over 90% of samples achieved a high LLM validation confidence score (>= 0.85) from Nemotron semantic evaluation.

### Insights & Next Steps
- **Expand Under-represented Categories**: Mine additional samples for low-volume categories like `security_vulnerability` (2), `performance_issue` (2), and `off_by_one` (2) to balance category distributions.
- **Increase TypeScript Ratio**: Target TS-heavy repositories to balance language parity for JS vs TS code representation.